test/Microsoft.Azure.Devices.Edge.Test.Common/Runtime.cs

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Test.Common
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Test.Common.Config;
+    using Microsoft.Azure.Devices.Edge.Util;
+
+    using Registries = System.Collections.Generic.IEnumerable<(string address, string username, string password)>;
+
+    public class Runtime
+    {
+        readonly string agentImage;
+        readonly string deviceId;
+        readonly string hubImage;
+        readonly IotHub iotHub;
+        readonly bool optimizeForPerformance;
+        readonly Option<Uri> proxy;
+        readonly Registries registries;
+
+        public Runtime(string deviceId, string agentImage, string hubImage, Option<Uri> proxy, Registries registries, bool optimizeForPerformance, IotHub iotHub)
+        {
+            this.agentImage = agentImage;
+            this.deviceId = deviceId;
+            this.hubImage = hubImage;
+            this.iotHub = iotHub;
+            this.optimizeForPerformance = optimizeForPerformance;
+            this.proxy = proxy;
+            this.registries = registries;
+        }
+
+        // Returns a DateTime representing the moment the configuration was sent to IoT Hub. Some
+        // of the test modules begin sending events as soon as they are launched, so this timestamp
+        // can be used as a reasonable starting point when listening for events on the IoT hub's
+        // Event Hub-compatible endpoint.
+        public async Task<DateTime> DeployConfigurationAsync(Action<EdgeConfigBuilder> withConfig, CancellationToken token)
+        {
+            var builder = new EdgeConfigBuilder(this.deviceId);
+            builder.AddRegistryCredentials(this.registries);
+            builder.AddEdgeAgent(this.agentImage).WithProxy(this.proxy);
+            builder.AddEdgeHub(this.hubImage, this.optimizeForPerformance)
+                .WithEnvironment(new[] { ("RuntimeLogLevel", "debug") })
+                .WithProxy(this.proxy);
+
+            withConfig(builder);
+
+            DateTime deployTime = DateTime.Now;
```

Comment:
Move this line after DeployAsync call since deployTime is used for event verification, should it be better to set deployTime after DeployAsync is completed?

Reply:
It's possible (even if it's rather unlikely) that deployed modules could be up and running (and sending messages) before this test code starts executing again following `DeployAsync`. Marking the time before we even deploy is totally safe. Also, the event monitoring code adds an additional minute or two to account for clock drift, so precision isn't really important. We just want to make sure we can get the events we're looking for, while making a reasonable effort to eliminate noise.